MNT use check_scalar in SpectralBiClustering and SpectralCoClustering

creatornadiran opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
Reference Issue #20724
PR #20723
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.
Used check_scalar function instead of if-else blocks to validate parameters.

#### Any other comments?
Please let me know if there is a mistake or you have a suggestion.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

creatornadiran committed:
Used check_scalar to check parameters.

glemaitre commented on sklearn/cluster/_bicluster.py:
```diff
@@ -104,14 +106,22 @@ def __init__(
+        "n_clusters": { "target_type": numbers.Integral,"min_val": 1, "max_val": n_samples},
+        "n_init": { "target_type": numbers.Integral,"min_val": 1 },
+        }
+        for scalar_name in scalars_checks:
```
You can make 2 calls and not use a loop.

glemaitre commented on sklearn/cluster/_bicluster.py:
```diff
@@ -104,14 +106,22 @@ def __init__(
             )
+        scalars_checks = {
+        "n_clusters": { "target_type": numbers.Integral,"min_val": 1, "max_val": n_samples},
+        "n_init": { "target_type": numbers.Integral,"min_val": 1 },
```
There is some PEP8 issue here

glemaitre commented on sklearn/cluster/_bicluster.py:
```diff
@@ -494,22 +504,14 @@ def _check_parameters(self):
+        "n_components": { "target_type": numbers.Integral,"min_val": 1},
+        "n_best": { "target_type": numbers.Integral,"min_val": 1, "max_val": self.n_components },
+        }
+        for scalar_name in scalars_checks:
```
same here

creatornadiran committed:
Update _bicluster.py

PIP8 issues are fixed and loops are removed

creatornadiran commented:
Thanks for the reviews. I made the changes. If I'm still missing something, please let me know.

glemaitre commented:
The CIs are failing. Can you look at the reported error. You should be able to reproduce the error locally by running the associated tests of the scikit-learn estimator that you modified.

glemaitre commented:
I am posting the patch that would be required to fix most probably the CI with some additional changes that I would have asked later on with a good review:

```diff
diff --git a/sklearn/cluster/_bicluster.py b/sklearn/cluster/_bicluster.py
index 849afd6cf5..5bb8b95cc6 100644
--- a/sklearn/cluster/_bicluster.py
+++ b/sklearn/cluster/_bicluster.py
@@ -106,7 +106,7 @@ class BaseSpectral(BiclusterMixin, BaseEstimator, metaclass=ABCMeta):
         self.n_init = n_init
         self.random_state = random_state
 
-    def _check_parameters(self, n_samples):
+    def _check_parameters(self, n_samples, n_features):
         legal_svd_methods = ("randomized", "arpack")
         if self.svd_method not in legal_svd_methods:
             raise ValueError(
@@ -114,8 +114,13 @@ class BaseSpectral(BiclusterMixin, BaseEstimator, metaclass=ABCMeta):
                     self.svd_method, legal_svd_methods
                 )
             )
-        check_scalar(self.n_clusters, "n_clusters", target_type=numbers.Integral, min_val=1, max_val=n_samples)
-        check_scalar(self.n_init, "n_init", target_type=numbers.Integral, min_val=1)
+        check_scalar(
+            self.n_init,
+            "n_init",
+            target_type=numbers.Integral,
+            min_val=1,
+            include_boundaries="left",
+        )
 
     def fit(self, X, y=None):
         """Creates a biclustering for X.
@@ -128,7 +133,7 @@ class BaseSpectral(BiclusterMixin, BaseEstimator, metaclass=ABCMeta):
 
         """
         X = self._validate_data(X, accept_sparse="csr", dtype=np.float64)
-        self._check_parameters(X.shape[0])
+        self._check_parameters(*X.shape)
         self._fit(X)
         return self
 
@@ -326,6 +331,17 @@ class SpectralCoclustering(BaseSpectral):
             n_clusters, svd_method, n_svd_vecs, mini_batch, init, n_init, random_state
         )
 
+    def _check_parameters(self, n_samples, n_features):
+        super()._check_parameters(n_samples, n_features)
+        check_scalar(
+            self.n_clusters,
+            "n_clusters",
+            target_type=numbers.Integral,
+            min_val=1,
+            max_val=n_samples,
+            include_boundaries="both",
+        )
+
     def _fit(self, X):
         normalized_data, row_diag, col_diag = _scale_normalize(X)
         n_sv = 1 + int(np.ceil(np.log2(self.n_clusters)))
@@ -487,8 +503,8 @@ class SpectralBiclustering(BaseSpectral):
         self.n_components = n_components
         self.n_best = n_best
 
-    def _check_parameters(self, n_sample):
-        super()._check_parameters()
+    def _check_parameters(self, n_samples, n_features):
+        super()._check_parameters(n_samples, n_features)
         legal_methods = ("bistochastic", "scale", "log")
         if self.method not in legal_methods:
             raise ValueError(
@@ -496,22 +512,60 @@ class SpectralBiclustering(BaseSpectral):
                     self.method, legal_methods
                 )
             )
-        try:
-            int(self.n_clusters)
-        except TypeError:
+
+        n_clusters_type_error = (
+            f"Incorrect parameter n_clusters has value: {self.n_clusters}. It "
+            "should either be a single integer or an iterable with two "
+            "integers: (n_row_clusters, n_column_clusters)"
+        )
+        if isinstance(self.n_clusters, numbers.Integral):
+            check_scalar(
+                self.n_clusters,
+                "n_clusters",
+                target_type=numbers.Integral,
+                min_val=1,
+                max_val=n_samples,
+                include_boundaries="both",
+            )
+        elif isinstance(self.n_clusters, tuple):
             try:
-                r, c = self.n_clusters
-                int(r)
-                int(c)
-            except (ValueError, TypeError) as e:
-                raise ValueError(
-                    "Incorrect parameter n_clusters has value:"
-                    " {}. It should either be a single integer"
-                    " or an iterable with two integers:"
-                    " (n_row_clusters, n_column_clusters)"
-                ) from e
-        check_scalar(self.n_components, "n_components", target_type=numbers.Integral, min_val=1)
-        check_scalar(self.n_best, "n_best", target_type=numbers.Integral, min_val=1, max_val=self.n_components)
+                rows, columns = self.n_clusters
+            except ValueError as e:
+                raise ValueError(n_clusters_type_error) from e
+            check_scalar(
+                rows,
+                "n_rows from n_clusters",
+                target_type=numbers.Integral,
+                min_val=1,
+                max_val=n_samples,
+                include_boundaries="both",
+            )
+            check_scalar(
+                columns,
+                "n_columns from n_clusters",
+                target_type=numbers.Integral,
+                min_val=1,
+                max_val=n_features,
+                include_boundaries="both",
+            )
+        else:
+            raise TypeError(n_clusters_type_error)
+
+        check_scalar(
+            self.n_components,
+            "n_components",
+            target_type=numbers.Integral,
+            min_val=1,
+            include_boundaries="left",
+        )
+        check_scalar(
+            self.n_best,
+            "n_best",
+            target_type=numbers.Integral,
+            min_val=1,
+            max_val=self.n_components,
+            include_boundaries="both",
+        )
 
     def _fit(self, X):
         n_sv = self.n_components
diff --git a/sklearn/cluster/tests/test_bicluster.py b/sklearn/cluster/tests/test_bicluster.py
index ba6d91a537..0eb6a6805e 100644
--- a/sklearn/cluster/tests/test_bicluster.py
+++ b/sklearn/cluster/tests/test_bicluster.py
@@ -208,23 +208,36 @@ def test_perfect_checkerboard():
 
 
 @pytest.mark.parametrize(
-    "args",
+    "args, err_type, err_msg",
     [
-        {"n_clusters": (3, 3, 3)},
-        {"n_clusters": "abc"},
-        {"n_clusters": (3, "abc")},
-        {"method": "unknown"},
-        {"n_components": 0},
-        {"n_best": 0},
-        {"svd_method": "unknown"},
-        {"n_components": 3, "n_best": 4},
+        (
+            {"n_clusters": (3, 3, 3)},
+            ValueError,
+            r"Incorrect parameter n_clusters has value: \(3, 3, 3\)",
+        ),
+        (
+            {"n_clusters": "abc"},
+            TypeError,
+            "Incorrect parameter n_clusters has value: abc",
+        ),
+        (
+            {"n_clusters": (3, "abc")},
+            TypeError,
+            "n_columns from n_clusters must be an instance of <class"
+            " 'numbers.Integral'>, not <class 'str'>.",
+        ),
+        ({"method": "unknown"}, ValueError, "Unknown method: 'unknown'"),
+        ({"n_components": 0}, ValueError, "n_components == 0, must be >= 1."),
+        ({"n_best": 0}, ValueError, "n_best == 0, must be >= 1."),
+        ({"svd_method": "unknown"}, ValueError, "Unknown SVD method: 'unknown'"),
+        ({"n_components": 3, "n_best": 4}, ValueError, "n_best == 4, must be <= 3."),
     ],
 )
-def test_errors(args):
+def test_errors(args, err_type, err_msg):
     data = np.arange(25).reshape((5, 5))
 
     model = SpectralBiclustering(**args)
-    with pytest.raises(ValueError):
+    with pytest.raises(err_type, match=err_msg):
         model.fit(data)
 
 
```

In short, this patch does:

- specialize each `Spectral` classes for the `n_clusters` check
- rework the type of error raised for `n_clusters` => raise a proper `TypeError` when the type is not the one expected
- improve the tests by checking the error message raised as well as the type of error

@creatornadiran Could you apply these changes such that I can make another round of reviews with these changes.

glemaitre reviewed (changes requested):
Adding the "Request changes" to see that we already reviewed this PR.

glemaitre commented:
`n_jobs` should also be checked.

creatornadiran committed:
Update _bicluster.py

creatornadiran commented:
> `n_jobs` should also be checked.

Wouldn't it be a bit difficult to check n_jobs with check_scalar?

creatornadiran commented:
CI test was giving the error due to PIP8 long line, I fixed it. Now passed the CI test.

creatornadiran committed:
Update _bicluster.py

creatornadiran committed:
Update _bicluster.py

creatornadiran committed:
Create _bicluster.py

I have activated the parameter checks

creatornadiran committed:
Update _bicluster.py

creatornadiran committed:
Update _bicluster.py

creatornadiran committed:
Update _bicluster.py

creatornadiran committed:
Update _bicluster.py

creatornadiran committed:
Update _bicluster.py

creatornadiran committed:
Update _bicluster.py

creatornadiran committed:
Update _bicluster.py

creatornadiran committed:
Update _bicluster.py

creatornadiran committed:
Update _bicluster.py

creatornadiran committed:
Update _bicluster.py

glemaitre commented:
> Wouldn't it be a bit difficult to check n_jobs with check_scalar?

```python
if self.n_jobs is not None:
    check_scalar(self.n_jobs, "n_jobs", numbers.Integral)
```

This should be enough

glemaitre commented:
Be aware that the CIs are failing. You should check the logs.

creatornadiran committed:
Update _bicluster.py

glemaitre commented:
Be sure to apply `black` on your file otherwise the linting CI will fail. The best is to install `pre-commit` as mentioned in the [contributing guide](https://scikit-learn.org/dev/developers/contributing.html#how-to-contribute) (item 9.) It will reformat the code for you before committing.

creatornadiran commented:
> > Wouldn't it be a bit difficult to check n_jobs with check_scalar?
> 
> ```python
> if self.n_jobs is not None:
>     check_scalar(self.n_jobs, "n_jobs", numbers.Integral)
> ```
> 
> This should be enough

I got "'SpectralCoclustering' object has no attribute 'n_jobs'" error.

creatornadiran commented:
> Be sure to apply `black` on your file otherwise the linting CI will fail. The best is to install `pre-commit` as mentioned in the [contributing guide](https://scikit-learn.org/dev/developers/contributing.html#how-to-contribute) (item 9.) It will reformat the code for you before committing.

I'm checking logs and there is no lining error but always linux-arm64 Cl is failing. I made too many commit but I will succeed i believe.  :relieved:

creatornadiran committed:
Update _bicluster.py

creatornadiran committed:
Update _bicluster.py

creatornadiran committed:
Update _bicluster.py

creatornadiran committed:
Update _bicluster.py

creatornadiran committed:
Update _bicluster.py

creatornadiran commented:
Finally all CLs are passed.

creatornadiran commented:
I added _check_parameters to SpectralCoclustering class is it OK?

creatornadiran commented on sklearn/cluster/_bicluster.py:
```diff
@@ -316,6 +321,20 @@ def __init__(
             n_clusters, svd_method, n_svd_vecs, mini_batch, init, n_init, random_state
         )
 
+    def _check_parameters(self, n_samples):
+        super()._check_parameters()
+        legal_methods = ('randomized', 'arpack')
+        if self.svd_method not in legal_methods:
+            raise ValueError(
+                "Unknown method: '{0}'. method must be one of {1}.".format(
+                    self.svd_method, legal_methods
+                )
+            )
+        check_scalar(self.n_clusters, "n_clusters",
+                     target_type=(numbers.Integral),
+                     min_val=1,
+                     max_val=n_samples)
+
```
I added _check_parameters to SpectralCoclustering class is it OK?

creatornadiran committed:
some change in spelling

thomasjpfan commented on sklearn/cluster/_bicluster.py:
```diff
@@ -124,7 +129,7 @@ def fit(self, X, y=None):
         """
         X = self._validate_data(X, accept_sparse="csr", dtype=np.float64)
-        self._check_parameters()
+        self._check_parameters(X.shape[0])
```
This now depends on child classes to define a `_check_parameters` that accepts `n_samples`, even tho the `BaseSpectral._check_parameters` does not.

May we update `BaseSpectral._check_parameters` to accept `n_samples` to be consistent?

thomasjpfan commented on sklearn/cluster/_bicluster.py:
```diff
@@ -316,6 +321,20 @@ def __init__(
 
+    def _check_parameters(self, n_samples):
+        super()._check_parameters()
+        legal_methods = ('randomized', 'arpack')
+        if self.svd_method not in legal_methods:
+            raise ValueError(
+                "Unknown method: '{0}'. method must be one of {1}.".format(
+                    self.svd_method, legal_methods
+                )
+            )
```
The call to `super()._check_parameters` will check `svd_method`.

```suggestion
```

thomasjpfan reviewed (commented):
Thank you for the PR @creatornadiran !

creatornadiran commented:
> Thank you for the PR @creatornadiran !

Thanks for the reply!

 I missed that the `super()._check_parameters` is already checking `svd_method`. I deleted. 

But  I can't get why should I add `n_samples` parameter to `BaseSpectral._check_parameters`. `n_samples` is the for check `n_cluster`'s max value but `n_cluster` checking from parent `_check_parameters` already.

creatornadiran committed:
Update sklearn/cluster/_bicluster.py

super()._check_parameters is already checking svd_method so no need to this block of code

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

thomasjpfan commented:
> But I can't get why should I add n_samples parameter to BaseSpectral._check_parameters. n_samples is the for check n_cluster's max value but n_cluster checking from child _check_parameters already.

`BaseSpectral._check_parameters` will not actually use `n_samples` for checking. It's more to make the API consistent.

Currently, `BaseSpectral.fit` calls `self._check_parameters(X.shape[0])`, but it's own `_check_parameters` does not accept `n_samples`. `BaseSpectral.fit` depends on a subclass to **override and change the signature** of `_check_parameters`, which feels non-intuitive.

creatornadiran committed:
Update _bicluster.py

creatornadiran commented:
I understand now thanks. I made the changes.

thomasjpfan commented on sklearn/cluster/_bicluster.py:
```diff
@@ -104,14 +106,17 @@ def __init__(
                 )
             )
+        check_scalar(self.n_init, "n_init", target_type=numbers.Integral, min_val=1)
+        if hasattr(self, 'n_jobs') and self.n_jobs is not None:
+            check_scalar(self.n_jobs, "n_jobs", numbers.Integral)
```
I do not think any subclass of `BaseSpectral` define `n_jobs`:

```suggestion
```

(`SpectralCoclustering` use to define it, but it has been removed in 1.0)

thomasjpfan reviewed (approved):
Small comment regarding `n_jobs`, otherwise LGTM!

creatornadiran committed:
Update sklearn/cluster/_bicluster.py

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

creatornadiran commented:
@glemaitre recommended me to check for `n_jobs`. But I think I used it in the wrong class.

thomasjpfan reviewed (commented):
There is a linting error, which can be fixed by running `black .`

thomasjpfan reviewed (approved):
LGTM

glemaitre reviewed (commented):
In general, we need to go back to the test and check for the error message that is issued.
By doing so, I already found that we have some bugs (at least we don't raise the proper error message).

thomasjpfan commented:
> ImportError: numpy is not installed. scikit-learn requires numpy >= 1.14.6. Installation instructions are available on the scikit-learn website: http://scikit-learn.org/stable/install.html from ci/circleci: linux-arm64

Syncing with main should resolve this issue.

creatornadiran committed:
runned black

creatornadiran committed:
necessary changes

creatornadiran committed:
test fixed

glemaitre commented on sklearn/cluster/tests/test_bicluster.py:
```diff
@@ -210,6 +210,7 @@ def test_perfect_checkerboard():
 @pytest.mark.parametrize(
     "args",
     [
+        {"n_init": 0},
```
In the `pytest.raises` context manager, we need to match the error message with `match=err_msg`. `err_msg` will be in the parametrisation.

glemaitre replied:
We should as well check that we trigger the `TypeError` by passing another type than the one expected.

creatornadiran committed:
bicluster_test updated

creatornadiran committed:
test_bicluster fixed

creatornadiran commented:
Updated, could you check again?

glemaitre commented on sklearn/cluster/tests/test_bicluster.py:
```diff
@@ -208,26 +208,43 @@ def test_perfect_checkerboard():
     ],
 )
-def test_errors(args):
+def test_value_errors(args1):
```
We don't need to test for the `ValueError` and `TypeError`. It can be part of the parametrization:

```python
@pytest.mark.parametrize(
    "params, type_err, err_msg",
    [
        ({"n_init": 0}, ValueError, "n_init == 0, must be >= 1."),
        ({"n_init": 1.5}, TypeError, "n_init must be an instance of"),
        (
            {"n_clusters": (3, 3, 3)},
            ValueError,
            r"Incorrect parameter n_clusters has value: \(3, 3, 3\)",
        ),
        (
            {"n_clusters": (6, 6)},
            ValueError,
            r"Incorrect parameter n_clusters has value: \(6, 6\)",
        ),
        (
            {"n_clusters": "abc"},
            ValueError,
            "Incorrect parameter n_clusters has value: abc",
        ),
        (
            {"n_clusters": (3, "abc")},
            ValueError,
            r"Incorrect parameter n_clusters has value: \(3, 'abc'\)",
        ),
        (
            {"n_clusters": ("abc", 3)},
            ValueError,
            r"Incorrect parameter n_clusters has value: \('abc', 3\)",
        ),
        ({"method": "unknown"}, ValueError, "Unknown method: 'unknown'"),
        ({"n_components": 0}, ValueError, "n_components == 0, must be >= 1."),
        ({"n_components": 1.5}, TypeError, "n_components must be an instance of"),
        ({"n_best": 0}, ValueError, "n_best == 0, must be >= 1."),
        ({"n_best": 1.5}, TypeError, "n_best must be an instance of"),
        ({"svd_method": "unknown"}, ValueError, "Unknown SVD method: 'unknown'"),
        ({"n_components": 3, "n_best": 4}, ValueError, "n_best == 4, must be <= 3."),
    ],
)
def test_spectalbiclustering_parameter_validation(params, type_err, err_msg):
    """Check parameters validation in `SpectralBiClustering`"""
    data = np.arange(25).reshape((5, 5))
    model = SpectralBiclustering(**params)
    with pytest.raises(type_err, match=err_msg):
        model.fit(data)
```

I don't see any test for `SpectralCoClustering`. We should have a similar test, however it should be limited to the available parameters.

creatornadiran committed:
test_bicluster.py fixed

creatornadiran committed:
test_bicluster refixed

creatornadiran committed:
test_bicluster

creatornadiran committed:
tests/test_bicluster fixed

creatornadiran commented:
Got these errors:
`


FAILED cluster/tests/test_bicluster.py::test_spectalcoclustering_parameter_validation[params2-ValueError-Incorrect parameter n_clusters has value: \\(3, 3, 3\\)]


FAILED cluster/tests/test_bicluster.py::test_spectalcoclustering_parameter_validation[params4-ValueError-Incorrect parameter n_clusters has value: abc]


FAILED cluster/tests/test_bicluster.py::test_spectalcoclustering_parameter_validation[params6-ValueError-Incorrect parameter n_clusters has value: \\('abc', 3\\)]


FAILED cluster/tests/test_bicluster.py::test_spectalbiclustering_parameter_validation[params1-TypeError-n_init must be an instance of integer]


FAILED cluster/tests/test_bicluster.py::test_spectalbiclustering_parameter_validation[params3-ValueError-n_clusters can not bigger than n_samples: n_clusters[1] == 6, must be <= 3]


FAILED cluster/tests/test_bicluster.py::test_spectalbiclustering_parameter_validation[params15-ValueError-n_clusters == 4, must be <= 3.]


FAILED cluster/tests/test_bicluster.py::test_spectalcoclustering_parameter_validation[params3-ValueError-Incorrect parameter n_clusters has value: \\(6, 6\\)]


FAILED cluster/tests/test_bicluster.py::test_spectalcoclustering_parameter_validation[params5-ValueError-Incorrect parameter n_clusters has value: \\(3, 'abc'\\)]


FAILED cluster/tests/test_bicluster.py::test_spectalcoclustering_parameter_validation[params7-ValueError-n_clusters can not bigger than n_samples: n_clusters == 4, must be <= 3.]


FAILED cluster/tests/test_bicluster.py::test_spectalbiclustering_parameter_validation[params4-ValueError-n_clusters can not bigger than n_samples: n_clusters[0] == 6, must be <= 3.]
`

glemaitre commented:
@creatornadiran you have to check the reason why they are failing in the CIs.
Be aware that you can run the tests locally to understand what the reasons for the tests to fail.

glemaitre commented:
A couple of changes after merging `main` into your branch

creatornadiran commented:
> @creatornadiran you have to check the reason why they are failing in the CIs. Be aware that you can run the tests locally to understand what the reasons for the tests to fail.

All of the errors are caused by the parameters in the test functions. I have not written a test function before, I would be very happy if you could help me.

glemaitre reviewed (commented):
Ups I see that I forgot to send my comments yesterday. It should help.

creatornadiran committed:
Update test_bicluster.py

creatornadiran committed:
rebase and changes

creatornadiran committed:
black

creatornadiran committed:
error fix

creatornadiran committed:
error fix 2

creatornadiran commented:
Still getting 2 errors and don't understand why.
Error Lines:
`FAILED cluster/tests/test_bicluster.py::test_spectalcoclustering_parameter_validation[params2-ValueError-Incorrect parameter n_clusters has value: abc]`

`FAILED cluster/tests/test_bicluster.py::test_spectalbiclustering_parameter_validation[params1-TypeError-n_init must be an instance of integer]`

glemaitre committed:
Merge remote-tracking branch 'origin/main' into pr/creatornadiran/20817

glemaitre committed:
fix regex to match

glemaitre commented:
> Still getting 2 errors and don't understand why.

The string that you passed was not matching the error message. You can check the fix here: https://github.com/scikit-learn/scikit-learn/pull/20817/commits/cda955d4b82f4722f6c9fc58ff2af569c8f56fd2

glemaitre commented:
CIs are green. Let's merge then. Thanks @creatornadiran

creatornadiran commented:
> > Still getting 2 errors and don't understand why.
> 
> The string that you passed was not matching the error message. You can check the fix here: [cda955d](https://github.com/scikit-learn/scikit-learn/commit/cda955d4b82f4722f6c9fc58ff2af569c8f56fd2)

Wow, I didn't notice. Thanks for the fix.
